src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,416 @@
+            return new BraceCompletionResult(formattingChanges, newClosingPoint - 1);
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext context, bool supportsVirtualSpace, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            var closingPoint = context.ClosingPoint;
+            var openingPoint = context.OpeningPoint;
+            var documentSnapshotText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            // Create a new source text that is not based on an editor snapshot so we can take advantage
+            // of the multi-version text change merging in ChangedText instead of ChangedSnapshotText.
+            var originalDocumentText = SourceText.From(documentSnapshotText.ToString(), documentSnapshotText.Encoding, documentSnapshotText.ChecksumAlgorithm);
+
+            // check whether shape of the braces are what we support
+            // shape must be either "{|}" or "{ }". | is where caret is. otherwise, we don't do any special behavior
+            if (!ContainsOnlyWhitespace(originalDocumentText, openingPoint, closingPoint))
+            {
+                return null;
+            }
+
+            // Insert a new line between the braces.
+            var newLineEdit = new TextChange(new TextSpan(closingPoint - 1, 0), Environment.NewLine);
+            var textWithNewLine = originalDocumentText.WithChanges(newLineEdit);
+
+            // Modify the closing point location to adjust for the newly inserted line.
+            closingPoint += Environment.NewLine.Length;
+
+            // Format the text that contains the newly inserted line.
+            var (formattingChanges, newClosingPoint) = await FormatTrackingSpanAsync(
+                document.WithText(textWithNewLine),
+                openingPoint,
+                closingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: false,
+                rules: GetBraceFormattingRules(document),
+                cancellationToken).ConfigureAwait(false);
+            closingPoint = newClosingPoint;
+            var formattedText = textWithNewLine.WithChanges(formattingChanges);
+
+            // Get the empty line between the curly braces.
+            var desiredCaretLine = GetLineBetweenCurlys(closingPoint, formattedText);
+            Debug.Assert(desiredCaretLine.GetFirstNonWhitespacePosition() == null, "the line between the formatted braces is not empty");
+
+            if (supportsVirtualSpace)
+            {
+                var overallChanges = formattedText.GetTextChanges(originalDocumentText).ToImmutableArray();
+                return new BraceCompletionResult(overallChanges, desiredCaretLine.Start);
+            }
+            else
+            {
+                // The caller does not support virtual spaces, so we have to insert actual whitespace into the document.
```

Comment:
fwiw, i think this should potentially be hte callers responsibility.  but i'm ont going to push hard on that if it turns out to be extremely difficult to do.

but it does seem like this API should just return data.  So it woudl say: here are the text changes.  And after that text change, the caret should be on line 'X' at positoin 'Y'.  Host then says: sure, i'll place it htere with virtual space.  Or "can't do that.  i'll insert whitespcae".

The concept of 'virtual space' should not leak into 'Features'.

Reply:
this also feels like this solve the "i need to pass back (or merge) multiple set of textchangeranges"

this helper will always just give you the formatting changes, and the final desired line/col.  The higher up layer will be responsible for handling that.